controllers/PlagiarismWebhookHandler.inc.php

```diff
@@ -0,0 +1,179 @@
+
+		if ($payload->status !== 'COMPLETE') {
+			// If the status not `COMPLETE`, then it's `ERROR`
+			static::$_plugin->sendErrorMessage(
+				"Unable to schedule the similarity report generation for file id {$submissionFile->getId()} with error : " . __("plugins.generic.plagiarism.ithenticate.submission.error.{$payload->error_code}"),
+				$submissionFile->getData('submissionId')
+			);
+			return;
+		}
+		
+		if ((int)$submissionFile->getData('ithenticateSimilarityScheduled')) {
+			static::$_plugin->sendErrorMessage("Similarity report generation process has already been scheduled for iThenticate submission id {$payload->id}", $submissionFile->getData('submissionId'));
+			return;
+		}
+		
+		list($apiUrl, $apiKey) = static::$_plugin->getServiceAccess($context);
+		$ithenticate = static::$_plugin->initIthenticate($apiUrl, $apiKey);
+
+		$scheduleSimilarityReport = $ithenticate->scheduleSimilarityReportGenerationProcess(
+			$payload->id,
+			static::$_plugin->getSimilarityConfigSettings($context)
+		);
+
+		if (!$scheduleSimilarityReport) {
+			static::$_plugin->sendErrorMessage("Failed to schedule the similarity report generation process for iThenticate submission id {$payload->id}", $submissionFile->getData('submissionId'));
+			return;
+		}
+
+		$submissionFile->setData('ithenticateSimilarityScheduled', 1);
+		$submissionFileDao->updateObject($submissionFile);
+	}
+
+	/**
+	 * Store or Update the result of similarity check for a submission file at receiving
+	 * the webook event `SIMILARITY_COMPLETE` or `SIMILARITY_UPDATED`
+	 * 
+	 * @param Context 	$context 	The current context for which the webhook request has initiated
+	 * @param string 	$payload	The incoming request payload through webhook
+	 * @param string 	$event		The incoming webhook request event
+	 *
+	 * @return void
+	 */
+	protected function storeSimilarityScore($context, $payload, $event) {
+		$payload = json_decode($payload);
+
+		// we will not store similarity check result unless it has completed
+		if ($payload->status !== 'COMPLETE') {
+			return;
+		}
+
+		$ithenticateSubmission = $this->getIthenticateSubmission($payload->submission_id);
+
+		if (!$ithenticateSubmission) {
+			static::$_plugin->sendErrorMessage("Invalid iThenticate submission id {$payload->submission_id} given for webhook event {$event}");
+			return;
+		}
+
+		/** @var SubmissionFileDAO $submissionFileDao */
+		$submissionFileDao = DAORegistry::getDAO('SubmissionFileDAO');
+		$submissionFile = $submissionFileDao->getById($ithenticateSubmission->submission_file_id);
+		$submissionFile->setData('ithenticateSimilarityResult', json_encode($payload));
+		$submissionFileDao->updateObject($submissionFile);
+	}
+
+	/**
+	 * Get the row data as object from submission file settings table or null if none found
+	 * 
+	 * @param string 	$id 	The given iThenticate submission id in UUID format
+	 * @return object|null
+	 */
+	private function getIthenticateSubmission($id) {
```

Comment:
This is OK for 3.3.0 (and I suppose you might need to do something similar for 3.4.0), but for `main`, make sure you add the necessary filter functions to the submission file's collector class rather than coding SQL in the outside world.